[ADBDEV-2501] Wrong subplan order when a query has a partitioned table

dnskvlnk opened the pull request:
This problem occurs with Postgres Query Optimizer only, ORCA is not affected.

If we have a partitioned table and a nested correlated subquery, we can face an error
that would say something like that:

```
FATAL:  Unexpected internal error (list.c:395)
DETAIL:  FailedAssertion("!(n < list->length)", File: "list.c", Line: 395)
```
In a case when a query has a partitioned table and InitPlan can't be created,
GPDB will make copies of first level subplans by number of partitions, and
all more nested subplans will be added later. But all more nested subplans copies
will be added in a wrong order and that they will go in a flat plan list after thier
parent nodes, what will lead to a crash.

```
create table t1 (a int) partition by range (a) (start (1) end (3) every (1));
create table t2 (b int8) distributed by (b);
set optimizer = off;
explain select 1 from t1 where a <= (
    SELECT 1 FROM t2
    WHERE t2.b <= (SELECT 1 from t2 as t3 where t3.b = t2.b)
    and t1.a = t2.b);
```
The intention of this patch is to create nested subplans right on the place. So,
by the time we get in nested_subplan_mutator, the plan already has dedicated
subplans for each partition. The intention of this part of
adjust_appendrel_attrs_mutator is to create a copy of the
most-closely-nested subplan and add it to subplans list. It's implied that
nested subplans of a copy will be added later in fixup_subplans, but it's the
cause of a crush: more nested subplans will be added after their parent subplans
in a flat list. The intension of this fix is to recursively traverse through
all the subplans and add them right on the spot. We make subplan copies
in nested_subplan_mutator and expect that when we get to fixup_subplans,
all the nested subplans are already initialized and fixup_subplans
doens't need to make any copies anymore.

dnskvlnk committed:
added add_nested_subplans

By the time we get here, the plan already has dedicated subplans for each partition. The intention of this part of adjust_appendrel_attrs_mutator is to create a copy of the most-closely-nested subplan and add it to subplans list. It's implied that nested subplans of a copy will be added later in fixup_subplans, but it's the cause of a crush: more nested subplans will be added after their parent subplansin a flat list. The intension of this fix is to recursively traverse through all the subplans and add them right on the spot. This will make the order of subplans correct and will leave nothing to do for fixup_subplans.

dnskvlnk committed:
renamed add_nested_subplans to nested_subplans_mutator

Keeping names in accordance with the convention

dnskvlnk reviewed (commented):
adjust_appendrel_attrs_mutator recieves a plan, which has seperate nodes for each partition, but they have no nested subplans. adjust_appendrel_attrs_mutator recursively traverses plan tree, but stops when it finds Subplan node and makes a copy of first child node. When this function retuns, the plan remains incomlete. This fix makes it traverse the plan tree to the most nested suplan and make copies for all nested subplans.

Stolb27 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -1808,6 +1808,40 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
+
+			memcpy(newsubroot, planner_subplan_get_root(root, sp), sizeof(PlannerInfo));
+
+			plan_tree_walker(newsubplan, nested_subplans_mutator, context);
```
`plan_tree_walker` expects that context is a pointer to `plan_tree_base_prefix`-compatible struct ([here](https://github.com/arenadata/gpdb/blob/ADBDEV-2501/src/backend/optimizer/util/walkers.c#L439)). Can you explain why do you use `adjust_appendrel_attrs_context` instead?

dnskvlnk replied:
I've added a new struct - `nested_subplans_context`, that's a derivative of `plan_tree_base_prefix`

Stolb27 replied:
I think that defining struct name in both typedef and tag namespaces is more common pattern for gpdb code base.

dnskvlnk replied:
Can you explain what you mean about tag namespace? Is that a value for a `type` field of a `Node`?

Stolb27 replied:
`typedef struct TagNamespaceName {} TypedefNamespaceName`

dnskvlnk committed:
added nested_subplans_context and and nested_subplans wraparound

nested_subplans_context is a derivative of plan_tree_base_prefix and it's used by plan_tree_walker. nested_subplans is a wraparound for nested_subplan_mutator, it converts adjust_appendrel_attrs_context to nested_subplans_context.

Stolb27 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -1808,6 +1808,54 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
+	if (IsA(node, SubPlan)) {
+		SubPlan *sp = (SubPlan*) node;
+
+		if (!sp->is_initplan) {
```
@dnskvlnk, currently you don't recurse down into initplans. Do we really want to avoid such situations?
Otherwise, we could recurse down one time before `if (IsA(node, SubPlan)) {` condition.

dnskvlnk replied:
Originally, the planner didn't recurse into initplans because it didn't want to create any copies of subplans inside initplans, because initplans are executed independently. When we copied this logic, we didn't want this either, so the behavior when we avoid initplans is intentional.

Stolb27 replied:
It seems you are right and gpdb handle possible partition table scan inside initplans independently. So we don't need to recurse into them.

Stolb27 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -1808,6 +1808,54 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
+
+static bool
+nested_subplans(Node *node, adjust_appendrel_attrs_context *context) {
+	nested_subplans_context *new_context = (nested_subplans_context*)malloc(sizeof(nested_subplans_context));
+	new_context->root = context->root;
```
Maybe move this logic to `adjust_appendrel_attrs_mutator` function?
Don't use c allocator (malloc) without strict needless. Use palloc instead.

dnskvlnk replied:
I assume that the usage of wraparounds is common in gpdb, so the intention of nested_subplans is to hide the conversion from adjust_appendrel_attrs_mutator and keep it tidy. 
Regerding malloc, can you suggest a common allocator, because as I can see, all the allocators are type-specific?

Stolb27 replied:
I don't think that this walker could be used somewhere else.
Please read article about Postgres memory allocation: src/backend/utils/mmgr/README
In this case, we can put struct item on stack and avoid memory allocation at all.

dnskvlnk replied:
I swithced from malloc to palloc and used pfree after `nested_subplan_mutator`. I also removed a wraparound, and the allocation is now located in `adjust_appendrel_attrs_mutator`

Stolb27 replied:
Do we really need memory allocation?

> In this case, we can put struct item on stack and avoid memory allocation at all.

Why do you leave `base->node` undefined?

dnskvlnk replied:
> Why do you leave `base->node` undefined?

The `node` field can have different types and the name of this field doesn't reflect the type of its content. I added the `root` field to `nested_subplans_context` to keep it similar to `adjust_appendrel_attrs_mutator`.

Stolb27 replied:
> Structure plan_tree_base_prefix carries the appropriate pointer for GPDB's general plan tree walker/mutator framework.  **All users** of the framework **must prefix** their context structure with a plan_tree_base_prefix and **initialize** it appropriately.

according to the structure [comment](https://github.com/arenadata/gpdb/blob/619755cdcd8c197e9fb8339a5b33ad80e02f9995/src/include/optimizer/walkers.h#L16-L27). Of course, currently, you handle SubPlan nodes separately in your function. But agreements were invented to simplify code maintenance.

> The plan associated with a SubPlan is found in a list.  During planning this is in the global structure found through the root PlannerInfo.

It seems, that you duplicate entity for root.

dnskvlnk committed:
fixed nested_subplans_mutator to nested_subplan_mutator

dnskvlnk committed:
changed the order of nested_subplans and nested_subplan_mutator

dnskvlnk committed:
added regression tests

dnskvlnk committed:
Add nested_subplan_context

dnskvlnk committed:
Add nested_subplan_context allocation and release

dnskvlnk committed:
Switch from adjust_appendrel_attrs_context to nested_subplan_context

dnskvlnk committed:
Change nested_subplans_mutator to nested_subplan_mutator

dnskvlnk committed:
Merge with ADBDEV-2501-fix and replace with theirs

Stolb27 reviewed (changes requested):
1. Remove redundant memory allocation
2. Change code formatting according to nearest code (use pgindent)
3. Provide formatted commit message for commit with fixes.

Commit:
Change new_context allocation

Commit:
Add nested_subplan_mutator

By the time we get in nested_subplan_mutator, the plan already has dedicated
subplans for each partition. The intention of this part of
adjust_appendrel_attrs_mutator is to create a copy of the
most-closely-nested subplan and add it to subplans list. It's implied that
nested subplans of a copy will be added later in fixup_subplans, but it's the
cause of a crush: more nested subplans will be added after their parent subplans
in a flat list. The intension of this fix is to recursively traverse through
all the subplans and add them right on the spot. We make subplan copies
is nested_subplan_mutator and expect that when we get to fixup_subplans,
all the nested subplans are already initialized and fixup_subplans
doens't need to make any copies anymore.

InnerLife0 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -2059,28 +2097,10 @@ adjust_appendrel_attrs_mutator(Node *node,
-
-			memcpy(newsubroot, planner_subplan_get_root(root, sp), sizeof(PlannerInfo));
+		plan_tree_base_prefix new_context;
+		new_context.node = (Node*)context->root;
```
I suggest to run [pgindent](https://github.com/arenadata/gpdb/blob/6.20.4_arenadata34/src/tools/pgindent/README) on this file to keep code formatting clean.
Run it on one file - it will probably add a lot of formatting changes to whole file. No need to commit them all, just keep your new part of code modified by pgindent.

dnskvlnk replied:
I've fixed formatting of this line

InnerLife0 commented on src/test/regress/sql/partition_subquery.sql:
```diff
@@ -18,3 +18,14 @@ SELECT id FROM (
+explain select 1 from t1 where a <= (
+    SELECT 1 FROM t2
+    WHERE t2.b <= (SELECT 1 from t2 as t3 where t3.b = t2.b)
+    and t1.a = t2.b);
```
Don't forget we need trailing new line here.

dnskvlnk replied:
Fixed

InnerLife0 commented on src/backend/optimizer/prep/prepunion.c:
```diff
@@ -1808,6 +1808,44 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
+nested_subplan_mutator(Node *node, plan_tree_base_prefix *context) 
+{
+	if (node == NULL) 
+	{
```
Minor. We can ignore brackets here and save two lines :)

dnskvlnk replied:
This wal also fixed

Commit:
Fix formatting

Commit:
Add new line at the end of partition_subquery.sql

InnerLife0 commented:
A little messy regression test comment, but generally all is ok.
As Vasiliy mentioned, you'll need one formatted commit message, but it's probably better to modify it as a part of commits squashing when PR will be finally approved.

Commit:
Fix trailing spaces

Commit:
Add some data to make sure that the results are correct too
